### ISSUE TITLE
NPE in ClosureEventTriggeringInterceptor @ 207

### ISSUE THREAD

**Issue by kszymko** [grails/gorm-hibernate5#114]
This is a stopper to migrate from Grails 3.3.8 to Grails 3.3.9.

**Buggy code**

// The old state of the entity at the time it was last loaded from thedatabase; 
// can be **null** in the case of detached entities.
**Object[] oldState = hibernateEvent.getOldState();**

```
    private Map<String, Object> getModifiedPropertiesWithAutotimestamp(PreUpdateEvent hibernateEvent, ModificationTrackingEntityAccess entityAccess) {
        Map<String, Object> modifiedProperties = entityAccess.getModifiedProperties();

        EntityMetamodel entityMetamodel = hibernateEvent.getPersister().getEntityMetamodel();
        Integer dateCreatedIdx = entityMetamodel.getPropertyIndexOrNull(AutoTimestampEventListener.DATE_CREATED_PROPERTY);

        Object[] oldState = hibernateEvent.getOldState();
        Object[] state = hibernateEvent.getState();
        // Only for "dateCreated" property, "lastUpdated" is handled correctly
        if (dateCreatedIdx != null && oldState[dateCreatedIdx] != null && !oldState[dateCreatedIdx].equals(state[dateCreatedIdx])) {
            modifiedProperties.put(AutoTimestampEventListener.DATE_CREATED_PROPERTY, oldState[dateCreatedIdx]);
        }

        return modifiedProperties;
    }
```
Fix:

```
if (dateCreatedIdx != null && oldState[dateCreatedIdx] != null && !oldState[dateCreatedIdx].equals(state[dateCreatedIdx])) {
```

to

```
 if (dateCreatedIdx != null && oldState != null && oldState[dateCreatedIdx] != null && !oldState[dateCreatedIdx].equals(state[dateCreatedIdx])) {
```

**Stacktrace**

```
java.lang.NullPointerException: null
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.getModifiedPropertiesWithAutotimestamp(ClosureEventTriggeringInterceptor.java:207) ~[grails-datastore-gorm-hibernate-core-6.1.11.RELEASE.jar:?]
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.synchronizeHibernateState(ClosureEventTriggeringInterceptor.java:189) ~[grails-datastore-gorm-hibernate-core-6.1.11.RELEASE.jar:?]
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onPreUpdate(ClosureEventTriggeringInterceptor.java:250) ~[grails-datastore-gorm-hibernate-core-6.1.11.RELEASE.jar:?]
	at org.hibernate.action.internal.EntityUpdateAction.preUpdate(EntityUpdateAction.java:244) ~[hibernate-core-5.1.5.Final.jar:5.1.5.Final]
```

**Environment Information**
Operating System: Windows 10 x64
GORM Version: 6.1.11
Grails Version (if using Grails): 3.3.9
JDK Version: 8

**Comment by allnightlong** [comment]
Same issue here

**Comment by jascha84** [comment]
same here...
when will you fix this issue?

```
java.lang.NullPointerException
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.getModifiedPropertiesWithAutotimestamp(ClosureEventTriggeringInterceptor.java:207)
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.synchronizeHibernateState(ClosureEventTriggeringInterceptor.java:189)
	at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onPreUpdate(ClosureEventTriggeringInterceptor.java:250)
	at org.hibernate.action.internal.EntityUpdateAction.preUpdate(EntityUpdateAction.java:244)
	at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:118)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:600)
	at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:474)
	at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
	at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
	at org.hibernate.internal.SessionImpl.doFlush(SessionImpl.java:1437)
	at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1423)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.flushSession(AbstractHibernateGormInstanceApi.groovy:289)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave_closure3(AbstractHibernateGormInstanceApi.groovy:254)
	at groovy.lang.Closure.call(Closure.java:418)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:299)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:243)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:117)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave(AbstractHibernateGormInstanceApi.groovy:251)
	at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.save(AbstractHibernateGormInstanceApi.groovy:168)
	at org.grails.datastore.gorm.GormEntity$Trait$Helper.save(GormEntity.groovy:151)
```

**Comment by graemerocher** [comment]
We will get to it eventually, otherwise feel free to send a PR or contract OCI support for priority bug fixing https://objectcomputing.com/products/grails

**Comment by aswinigbc** [comment]
So do i need to upgrade hibernate plugin in grails 3.3.9 or .. Can someone elaborate the steps how to fix this issue in grails v3.3.9

 `Caused by: java.lang.NullPointerException: null
        at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.getModifiedPropertiesWithAutotimestamp(ClosureEventTriggeringInterceptor.java:207)
        at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.synchronizeHibernateState(ClosureEventTriggeringInterceptor.java:189)
        at org.grails.orm.hibernate.support.ClosureEventTriggeringInterceptor.onPreUpdate(ClosureEventTriggeringInterceptor.java:250)
        at org.hibernate.action.internal.EntityUpdateAction.preUpdate(EntityUpdateAction.java:244)
        at org.hibernate.action.internal.EntityUpdateAction.execute(EntityUpdateAction.java:118)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:589)
        at org.hibernate.engine.spi.ActionQueue.executeActions(ActionQueue.java:463)
        at org.hibernate.event.internal.AbstractFlushingEventListener.performExecutions(AbstractFlushingEventListener.java:337)
        at org.hibernate.event.internal.DefaultFlushEventListener.onFlush(DefaultFlushEventListener.java:39)
        at org.hibernate.internal.SessionImpl.flush(SessionImpl.java:1295)
        at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.flushSession(AbstractHibernateGormInstanceApi.groovy:289)
        at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi$_performSave_closure3.doCall(AbstractHibernateGormInstanceApi.groovy:254)
        at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:299)
        at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:243)
        at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:117)
        at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.performSave(AbstractHibernateGormInstanceApi.groovy:251)
        at org.grails.orm.hibernate.AbstractHibernateGormInstanceApi.save(AbstractHibernateGormInstanceApi.groovy:168)
        at org.grails.datastore.gorm.GormEntity$Trait$Helper.save(GormEntity.groovy:151)` 

Thanks

**Comment by HelainSchoonjans** [comment]
A workaround that worked for me was to do the operation in the context of a new hibernate session and ensure the domain is not detached:

`MyDomain.withNewSession {
    myDomainInstance.attach()
    myDomainInstance.field = 'new value'
    myDomainInstance.save()
}`
